tests/thread/main.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2025 Yury Gribov
+ *
+ * The MIT License (MIT)
+ * 
+ * Use of this source code is governed by MIT license that can be
+ * found in the LICENSE.txt file.
+ */
+
+#include <dlfcn.h>
+#include <stdio.h>
+#include <assert.h>
+#include <stdlib.h>
+
+#include <pthread.h>
+#include <unistd.h>
+
+#include "interposed.h"
+
+void *run(void *arg_) {
+  int *arg = (int *)arg_;
+
+  int (*foo)(int);
+  switch(*arg % 10) {
+    case 0:
+      foo = foo0;
+      break;
+    case 1:
+      foo = foo1;
+      break;
+    case 2:
+      foo = foo2;
+      break;
+    case 3:
+      foo = foo3;
+      break;
+    case 4:
+      foo = foo4;
+      break;
+    case 5:
+      foo = foo5;
+      break;
+    case 6:
+      foo = foo6;
+      break;
+    case 7:
+      foo = foo7;
+      break;
+    case 8:
+      foo = foo8;
+      break;
+    case 9:
+      foo = foo9;
+      break;
+    default:
+      abort();
+  }
+
+  *arg = foo(*arg);
+
+  return 0;
+}
+
+#define N 128
+
+int args[N];
+pthread_t tids[N];
+
+int main() {
+  int exp = 0;
+  for (int i = 0; i < N; ++i) {
+    args[i] = i;
+    exp += i + (i % 10);
+  }
+
+  for (int i = 0; i < N; ++i) {
+    if (0 != pthread_create(&tids[i], 0, run, &args[i]))
```

Comment:
The `run` is very short-lived. It's possible for the launched thread to run and finish before the main thread manages to launch the next one.

A more robust way to ensure a race would be to make all threads block on something locked by the main thread, and then unlock all of them at once. E.g use `pthread_barrier_wait` in `run` to wait until the last thread has been launched.